BrainPortal/lib/authenticated_system.rb

```diff
@@ -126,11 +140,25 @@ def redirect_back_or_default(default)
     # available as ActionView helper methods.
     def self.included(base)
       base.send :helper_method, :current_user, :logged_in?
+      base.class_eval do
+        before_action :extract_api_token
+      end
     end
 
     # Called from #current_user.  First attempt to login by the user id stored in the session.
-    def login_from_session
-      self.current_user = User.find_by_id(session[:user_id]) if session[:user_id]
+    def user_from_session
+      User.find_by_id(session[:user_id]) if session[:user_id]
+    end
+
+    # For API calls. A +cbrain_api_token+ is expected in the params.
+    # If the token is valid, the cbrain_session object will contain
+    # a proper LargeSessionInfo object from which the token's user
+    # can be found.
+    def user_from_api_token
+      return nil unless @cbrain_api_token
+      user = cbrain_session.user_for_cbrain_api(@cbrain_api_token)
+      #@cbrain_api_token = nil unless user # invalid, just reset it
```

Comment:
Why did you keep this comment?

Reply:
Because I am not yet sure if, for the general behavior of CBRAIN, I want to zap the token if it's invalid, or keep it around.